src/promql/src/extension_plan/range_manipulate.rs

```diff
@@ -0,0 +1,504 @@
+    end: Millisecond,
+    interval: Millisecond,
+    range: Millisecond,
+    time_index: usize,
+    value_columns: Vec<usize>,
+
+    output_schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    metric: BaselineMetrics,
+}
+
+impl RecordBatchStream for RangeManipulateStream {
+    fn schema(&self) -> SchemaRef {
+        self.output_schema.clone()
+    }
+}
+
+impl Stream for RangeManipulateStream {
+    type Item = ArrowResult<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let poll = match self.input.poll_next_unpin(cx) {
+            Poll::Ready(batch) => {
+                let _timer = self.metric.elapsed_compute().timer();
+                Poll::Ready(batch.map(|batch| batch.and_then(|batch| self.manipulate(batch))))
+            }
+            Poll::Pending => Poll::Pending,
+        };
+        self.metric.record_poll(poll)
+    }
+}
+
+impl RangeManipulateStream {
+    pub fn manipulate(&self, input: RecordBatch) -> ArrowResult<RecordBatch> {
+        let mut other_columns = (0..input.columns().len()).collect::<HashSet<_>>();
+        // calculate the range
+        let ranges = self.calculate_range(&input);
+        // transform columns
+        let mut new_columns = input.columns().to_vec();
+        for index in self.value_columns.iter().chain([self.time_index].iter()) {
+            other_columns.remove(index);
+            let column = input.column(*index);
+            let new_column = Arc::new(
+                RangeArray::from_ranges(column.clone(), ranges.clone())
+                    .map_err(|e| ArrowError::InvalidArgumentError(e.to_string()))?
+                    .into_dict(),
+            );
+            new_columns[*index] = new_column;
+        }
+
+        // truncate other columns
+        let take_indices = Int64Array::from(vec![0; ranges.len()]);
+        for index in other_columns.into_iter() {
+            new_columns[index] = compute::take(&input.column(index), &take_indices, None)?;
+        }
+
+        RecordBatch::try_new(self.output_schema.clone(), new_columns)
+    }
+
+    fn calculate_range(&self, input: &RecordBatch) -> Vec<(u32, u32)> {
+        let ts_column = input
+            .column(self.time_index)
+            .as_any()
+            .downcast_ref::<TimestampMillisecondArray>()
+            .unwrap();
+
+        let mut result = vec![];
+
+        for curr_ts in (self.start..=self.end).step_by(self.interval as _) {
+            let mut range_start = ts_column.len();
+            let mut range_end = 0;
+            for (index, ts) in ts_column.values().iter().enumerate() {
+                if ts + self.range >= curr_ts {
+                    range_start = range_start.min(index);
+                }
+                if *ts <= curr_ts {
+                    range_end = range_end.max(index);
+                }
+            }
+            result.push((range_start as _, (range_end + 1 - range_start) as _));
```

Comment:
What if no timestamp could be found in a range? Do we need to check `range_start <= range_end` here?

Reply:
return an empty range `(0,0)`, nice catch